### PR TITLE
fix: #692 Admin FOM should not be editable when commenting open.

### LINKED ISSUE #692: Disable FOM form modifications/updates when commenting period is open
**Describe the task**
Email from Julius: 
Image

Image
Image

**Acceptance Criteria**
- [ ] first
- [ ] second
- [ ] third

**Additional context**
- Add any other context about the task here.
- Or here

### PR DESCRIPTION
**Pull request by ianliuwk1019**
ref: #692 
Business requirement changes due to issue the licenses edit FOM during commenting open (description field) and it has legal consequence. New rule: All FOM fields should not be editable except attachments fields during commenting open. 

- Adjust fom-add-edit component so that fields are only new/editable during FOM initial state. On Commenting open and any other state they will be disabled. Some field are only editable during "create" (initial state).

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-43.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-43.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-43.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-43.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-43.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-43.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

### CODE REVIEW

File: api/src/main.ts [hunk 1]
```diff
@@ -189,4 +189,4 @@ if (process.argv.length > 2 &&
   standaloneRunTestDataMigrations();
 } else {
   startApi();
-} 
\ No newline at end of file
+}
```
Comment (ianliuwk1019) [inline api/src/main.ts:192]: Just dummy change for this, to trigger deployment correctly; can ignore this file..

File: admin/src/core/services/mock-user.ts [hunk 2]
```diff
@@ -53,5 +53,6 @@ export function getFakeAllAccessUser(): User {
   user.isForestClient = true;
   user.clientIds.push('00001011')
   user.clientIds.push('00001012');
+  user.clientIds.push('00132188');
```
Comment (ianliuwk1019) [inline admin/src/core/services/mock-user.ts:56]: Can ignore this file, only add a test client for Timber Sales later.

**Comment by basilv** [comment]
So my main concern is that we're changing a number of fields to be only editable in initial state when they were editable in multiple states before. So what if these fields still need to be editable when commenting is closed? I'm not saying they need to be, but it isn't clear what change we are actually making, versus just making fields non-editable during commenting-open.

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.html [hunk 3]
```diff
@@ -308,6 +395,15 @@ <h1 class="text-muted">{{isCreate ? 'Add New' : 'Edit'}} FOM {{isCreate ? '' : '
...
             />
+            <input
+              *ngIf="!isInitialState"
+              type="text"
+              disabled
+              class="form-control"
+              id="mgr-name-id"
+              name="bctsMgrName"
+              [value]="fg.get('bctsMgrName').value"
+            />
```
Comment (craigyu) [inline admin/src/app/foms/fom-add-edit/fom-add-edit.component.html]: might be a good idea to reduce duplication to use ternary in places like this, for example, the two <input> can be merged into one by doing 
```html
<input
  type="text"
  class="form-control"
  id="mgr-name-id"
  name="bctsMgrName"
  formControlName="bctsMgrName"
  [appFormControl]="fg.get('bctsMgrName')"
  [attr.disabled]="!isInitialState ? true : null"
  [value]="isInitialState ? null : fg.get('bctsMgrName').value"
/>
```

Although this is fine for a hotfix, to the tech debt it goes!

**Review by craigyu — APPROVED** [review]
I don't have things setup locally to test these but the code looks good!

**Comment by craigyu** [comment]
> So my main concern is that we're changing a number of fields to be only editable in initial state when they were editable in multiple states before. So what if these fields still need to be editable when commenting is closed? I'm not saying they need to be, but it isn't clear what change we are actually making, versus just making fields non-editable during commenting-open.

I had a state here to control the inputs depends on whether the form is in "comment open" state
https://github.com/bcgov/nr-fom/pull/693/commits/337ee0e4fa6063415d68ebced63f9ac42cbd4c78

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.html [hunk 4]
```diff
@@ -294,12 +379,14 @@ <h1 class="text-muted">{{isCreate ? 'Add New' : 'Edit'}} FOM {{isCreate ? '' : '
             </div>
           </div>
         </div>
+
         <div class="row" *ngIf="isHolderBctsManger()">
           <div class="form-group col">
-            <label [ngClass]="{'required': isHolderBctsManger()}">
+            <label [ngClass]="{'required': isInitialState}">
```
Comment (craigyu) [inline admin/src/app/foms/fom-add-edit/fom-add-edit.component.html]: can you replace `isHolderBctsManger()` with `isInitialState`?

**Comment by ianliuwk1019** [comment]
> So my main concern is that we're changing a number of fields to be only editable in initial state when they were editable in multiple states before. So what if these fields still need to be editable when commenting is closed? I'm not saying they need to be, but it isn't clear what change we are actually making, versus just making fields non-editable during commenting-open.

Thanks Basil! I adjusted the change, some fields are editable during commenting closed according to latest discussion from Julius.

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.html (hunk 3 above)
Comment (ianliuwk1019) [reply admin/src/app/foms/fom-add-edit/fom-add-edit.component.html]: Thanks Craig!! In this case due to many interaction from Angular, I consider it for a bit of clarity over conciseness.

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.html [hunk 5]
```diff
@@ -48,37 +49,52 @@ <h1 class="text-muted">{{isCreate ? 'Add New' : 'Edit'}} FOM {{isCreate ? '' : '
               [value]="originalProjectResponse.id"
             />
           </div>
+
+          <!-- Commenting Open Date: disabled when COMMENT_CLOSED -->
           <div class="form-group col-md">
             <label>Commenting Open Date</label>
-            <div class="input-group">
+              <div class="input-group" *ngIf="!isCommentingClosedState">
```
Comment (basilv) [inline admin/src/app/foms/fom-add-edit/fom-add-edit.component.html:56]: What happens if FOM is in state FINALIZED

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.html [hunk 6]
```diff
@@ -48,37 +49,52 @@ <h1 class="text-muted">{{isCreate ? 'Add New' : 'Edit'}} FOM {{isCreate ? '' : '
...
-                </button>
-              </div>
-            </div>
           </div>
+
+          <!-- Commenting Closed Date: disabled when COMMENT_CLOSED -->
           <div class="form-group col-md">
             <label>Commenting Closed Date</label>
-            <div class="input-group">
+            <div class="input-group" *ngIf="!isCommentingClosedState">
```
Comment (basilv) [inline admin/src/app/foms/fom-add-edit/fom-add-edit.component.html:97]: Should also be disabled when state = FINALIZED (everything should be read-only then I think).

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.ts [hunk 7]
```diff
@@ -169,7 +172,9 @@ export class FomAddEditComponent implements OnInit, AfterViewInit, OnDestroy {
         }
 
         this.forestClientSelect = this.originalProjectResponse.forestClient.id;
-
+        this.isInitialState = this.originalProjectResponse.workflowState.code === WorkflowStateEnum.Initial;
```
Comment (basilv) [inline admin/src/app/foms/fom-add-edit/fom-add-edit.component.ts:175]: Do we use isInitialState any more? If we do, then just resolve this comment.

**Review by basilv — COMMENTED** [review]
Have some questions...

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.html (hunk 5 above)
Comment (ianliuwk1019) [reply admin/src/app/foms/fom-add-edit/fom-add-edit.component.html:56]: Hi Basil, when state is finalized, the user has no option to come to this page (there is a condition for Edit button from FOM detail page).
Also the backend checked licenses only allowed for up to commenting closed to be editable, not for finalized..

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.ts (hunk 7 above)
Comment (ianliuwk1019) [reply admin/src/app/foms/fom-add-edit/fom-add-edit.component.ts:175]: I do use this initial condition on one place in html for "district".

File: admin/src/app/foms/fom-add-edit/fom-add-edit.component.html (hunk 6 above)
Comment (ianliuwk1019) [reply admin/src/app/foms/fom-add-edit/fom-add-edit.component.html:97]: Same as above in FINALIZED status, submitter cannot come to this page, and backend has a check, should be enough. Unless we like to give specific combination of status condition scenarios (we didn't have that previously).

**Review by basilv — APPROVED** [review]

**Review by DerekRoberts — COMMENTED** [review]
One more approval!

**Pull request merged**